functions/compare_classification_facts.pp

```diff
@@ -0,0 +1,39 @@
+# @summary Compare untrusted values from the classification fact to their trusted values.
+#
+# Compare untrusted values from the classification fact to their trusted values.
+#
+# @example
+#   $classification_calculated_trusted = $parsed_classification + {
+#     hostname => $hostname,
+#   }
+#   $classification_fact_differences = classification::compare_classification_facts($classification_calculated_trusted)
+#
+# @param classification_calculated_trusted A hash of facts to validate
+# @return [Array] An array of strings documenting any facts that don't match their trusted values
+function classification::compare_classification_facts(Hash $classification_calculated_trusted) >> Array {
+  $classification_fact_names = [
+    'hostname',
+    'version',
+    'group',
+    'function',
+    'number',
+    'number_string',
+    'context',
+    'stage',
+  ]
+
+  $classification_fact_differences = delete_undef_values($classification_fact_names.map |$fact_name| {
+    $untrusted_value = $facts['classification'][$fact_name]
+    $trusted_value = $classification_calculated_trusted[$fact_name]
+
+    # lint:ignore:80chars lint:ignore:140chars
+    if $untrusted_value != $trusted_value {
+      "    ${fact_name}: '${untrusted_value}' != '${trusted_value}'\n    ${fact_name} types: ${type($untrusted_value)} != ${type($trusted_value)}"
```

Comment:
I'd be inclined to get rid of the type check and convert these always to strings.

Reply:
This is what was in `dist/classification/manifests/init.p`:

~~~ puppet
  $differences = delete_undef_values($fact_names.map |$fact_name| {
    # INFC-17593: For some reason context is sometimes '' instead of undef.
    $untrusted_value = String($facts[$fact_name])
    $trusted_value = String($calculated_trusted[$fact_name])

    if $untrusted_value != $trusted_value {
      "    ${fact_name}: '${untrusted_value}' != '${trusted_value}'"
    } else {
      undef
    }
  })
~~~

Reply:
The type stuff is super helpful when it fails.

Reply:
It depends on if you always cast to string… if you do, it will always output String.

Reply:
ah, true... the extra info pre-dates the string cast.